### PR TITLE
Rule S4144: Methods should not have identical implementations

### LINKED ISSUE #610: Rule S4144: Methods should not have identical implementations
Implements [RSPEC-4144](https://jira.sonarsource.com/browse/RSPEC-4144)

### PR DESCRIPTION
**Pull request by Evangelink**
Fix #610

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+                !throwStatement.Expression.IsKind(SyntaxKind.ObjectCreationExpression))
+            {
+                return false;
+            }
+
+            var symbol = model.GetSymbolInfo(throwStatement.Expression).Symbol?.ContainingType;
+            return symbol.IsAny(ThrowStatementTypesToIgnore);
+        }
+
+        private class MethodDuplicates
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs]: MAJOR Remove the unused private type 'MethodDuplicates'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+                                    additionalLocations: new[] { method.Identifier.GetLocation() },
+                                    messageArgs: method.Identifier.ValueText));
+                            }
+                        }
+                    }
+                }, SyntaxKind.ClassDeclaration);
+        }
+
+        private bool AreDuplicates(MethodDeclarationSyntax firstMethod, MethodDeclarationSyntax secondMethod,
+            SemanticModel model)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs]: MAJOR Remove this unused method parameter 'model'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1172)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+                return false;
+            }
+
+            var symbol = model.GetSymbolInfo(throwStatement.Expression).Symbol?.ContainingType;
+            return symbol.IsAny(ThrowStatementTypesToIgnore);
+        }
+
+        private class MethodDuplicates
+        {
+            public MethodDuplicates(MethodDeclarationSyntax original, ICollection<MethodDeclarationSyntax> duplicates)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs]: MAJOR Remove the unused private constructor 'MethodDuplicates'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+            {
+                return firstMethod.Body.Statements.Count >= 2 &&
+                    firstMethod.Identifier.ValueText != secondMethod.Identifier.ValueText &&
+                    firstMethod.Body.IsEquivalentTo(secondMethod.Body, false);
+            }
+
+            return false;
+        }
+
+        private bool IsIgnoredThrowStatement(StatementSyntax statement, SemanticModel model)
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs]: MAJOR Remove the unused private method 'IsIgnoredThrowStatement'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)

**Comment by SonarTech** [comment]
SonarQube analysis reported 4 issues
* ![MAJOR][MAJOR] 4 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+    public sealed class MethodsShouldNotHaveIdenticalImplementations : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4144";
+        private const string MessageFormat = "Update this method so that its implementation is not identical to '{0}'.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        private static readonly ISet<KnownType> ThrowStatementTypesToIgnore = ImmutableHashSet.Create(
```
Comment (SonarTech) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs]: MAJOR Remove the unused private field 'ThrowStatementTypesToIgnore'. [rule](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=csharpsquid%3AS1144)

**Comment by SonarTech** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+                    var alreadyHandledMethods = new HashSet<MethodDeclarationSyntax>();
+
+                    foreach (var method in methods)
+                    {
+                        if (alreadyHandledMethods.Contains(method))
+                        {
+                            continue;
+                        }
+
+                        var duplicates = methods.Where(m => AreDuplicates(method, m)).ToList();
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs]: You could  probably use methods.Except(alreadyHandledMethods).Where(...)

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+                        if (duplicates.Count > 0)
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs]: Shouldn't we add method unconditionally? If this is the case, perhaps this check is not needed...

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        private bool AreDuplicates(MethodDeclarationSyntax firstMethod, MethodDeclarationSyntax secondMethod)
+        {
+            if (firstMethod == secondMethod)
+            {
+                return false;
+            }
+
+            if (firstMethod.Body != null &&
```
Comment (valhristov) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs]: can't we merge these ifs?

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs (hunk 7 above)
Comment (Evangelink) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/MethodsShouldNotHaveIdenticalImplementations.cs]: Not sure what you mean

**Pull request merged**